ert/data/record/_record.py

```diff
@@ -0,0 +1,234 @@
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
+
+    @staticmethod
+    def create(
+        spec_or_record: Union[
+            Dict[str, bytes], Dict[str, numerical_record_data], "Record"
+        ]
+    ) -> "Record":
+        if isinstance(spec_or_record, Record):
+            return spec_or_record
+        data = spec_or_record["data"]
+        if isinstance(data, bytes):
+            return BlobRecord(data)
+        else:
+            index = spec_or_record.get("index")
+            if not isinstance(index, int):
+                index = None
+            return NumericalRecord(data=data, index=index)
+
+
+class BlobRecord(Record):
+    def __init__(self, data: blob_record_data) -> None:
+        self._record_type = RecordType.BYTES
+
+        try:
+            self._data = self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+
+    @beartype
+    def _validate_data(self, data: blob_record_data) -> blob_record_data:
+        return data
+
+    @property
+    def data(self) -> blob_record_data:
+        return self._data
+
+    @property
+    def record_type(self) -> RecordType:
+        return RecordType.BYTES
+
+    def __eq__(self, o: object) -> bool:
+        if isinstance(o, BlobRecord):
+            return self.__dict__ == o.__dict__
+        return False
+
+
+class NumericalRecord(Record):
+    def __init__(
+        self, data: numerical_record_data, index: Optional[RecordIndex] = None
+    ) -> None:
+        if isinstance(data, (set, frozenset, deque, tuple)):
+            data = [val for _, val in enumerate(data)]
+
+        try:
+            self._data = self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+
+        self._type = self._infer_type()
+        self._index = self._validate_index(index)
+
+    @beartype
+    def _validate_data(self, data: numerical_record_data) -> numerical_record_data:
+        # beartype does not do deep validation on dicts, so we do non-nested validation
+        if isinstance(data, dict):
+            for key, val in data.items():
```

Comment:
Does this not loose the validation of homogeneous index and value type over the dictionary?

Reply:
Was the index constrained wrt. to homogeneity? Wouldn't `Tuple[Union[int, str], ...]` allow for `(1, 1.0)`?

Or are you referring to the `data` and `index` homogeneity constraint? It is checked in `_validate_index`

Reply:
You are right! I think we would benefit from swapping the Union and Tuple such that the index does indeed become homogeneous. I was planning to make the change in the type, but apparently never came around to do it. And when searching for an issue I can't find any. I remember discussing it with Peter in a PR he had a while back, but that seems to be it.

I'll write up an issue on this.

Reply:
https://github.com/equinor/ert/issues/2067